Add Batch Pool Container Registry configuration

ennui93 opened the pull request:
Add a `container_registries` block to the batch pool container configuration resource.

This setting is necessary for Azure Batch service Pools to support custom container images (i.e. from private container repositories).

katbyte commented on azurerm/data_source_batch_pool.go:
```diff
@@ -119,6 +119,16 @@ func dataSourceArmBatchPool() *schema.Resource {
+							Type:     schema.TypeSet,
+							Computed: true,
+							Elem: &schema.Schema{
+								Type: schema.TypeMap,
```
Instead of a map could we actually write out the schema?

ennui93 commented:
Hi @katbyte, thank you for your review. I have revamped the PR with the schema as requested. Please review and provide any pertinent feedback at your convenience.

katbyte commented on website/docs/r/batch_pool.html.markdown:
```diff
@@ -247,6 +256,18 @@ A `resource_file` block supports the following:
+
+A `container_registries` block supports the following:
+
+* `registry_server` - (Optional) The container registry URL. The default is "docker.io".
```
Could we add that this is a force new?
```suggestion
* `registry_server` - (Optional) The container registry URL. The default is "docker.io". Changing this forces a new resource to be created.
```

katbyte commented on website/docs/r/batch_pool.html.markdown:
```diff
@@ -247,6 +256,18 @@ A `resource_file` block supports the following:
+
+* `registry_server` - (Optional) The container registry URL. The default is "docker.io".
+
+* `user_name` - (Optional) The user name to log into the registry server.
```
And here

katbyte commented on website/docs/r/batch_pool.html.markdown:
```diff
@@ -247,6 +256,18 @@ A `resource_file` block supports the following:
+
+* `user_name` - (Optional) The user name to log into the registry server.
+
+* `password` - (Optional) The password to log into the registry server.
```
And here

katbyte commented on website/docs/r/batch_pool.html.markdown:
```diff
@@ -77,6 +77,13 @@ EOF
     type = "DockerCompatible"
+    container_registries = [
+      {
+        registry_server = "docker.io"
```
Could we align these assignments?

katbyte commented on azurerm/resource_arm_batch_pool.go:
```diff
@@ -124,6 +126,31 @@ func resourceArmBatchPool() *schema.Resource {
+										Type:     schema.TypeString,
+										Required: true,
+										ForceNew: true,
+									},
```
Could we validate this with at the very least `NoEmptyStrings`

katbyte commented on azurerm/resource_arm_batch_pool.go:
```diff
@@ -124,6 +126,31 @@ func resourceArmBatchPool() *schema.Resource {
+										Type:     schema.TypeString,
+										Required: true,
+										ForceNew: true,
+									},
```
Could we validate this with at the very least `NoEmptyStrings`

katbyte commented on azurerm/resource_arm_batch_pool.go:
```diff
@@ -124,6 +126,31 @@ func resourceArmBatchPool() *schema.Resource {
+										Type:      schema.TypeString,
+										Required:  true,
+										Sensitive: true,
+									},
```
Could we validate this with at the very least `NoEmptyStrings`

katbyte commented on azurerm/resource_arm_batch_pool_test.go:
```diff
@@ -311,7 +311,11 @@ func TestAccAzureRMBatchPool_container(t *testing.T) {
+					resource.TestCheckResourceAttr(resourceName, "container_configuration.0.container_registries.0.registry_server", "myContainerRegistry.azurecr.io"),
+					resource.TestCheckResourceAttr(resourceName, "container_configuration.0.container_registries.0.user_name", "myUserName"),
 				),
+				ExpectNonEmptyPlan: true,
```
We shouldn't need this.. i presume it is because we are reading in the password each time.

katbyte commented on website/docs/r/batch_pool.html.markdown:
```diff
@@ -247,6 +256,18 @@ A `resource_file` block supports the following:
+
+* `password` - (Optional) The password to log into the registry server.
+
+~> **NOTE:** Changes to the password field will not trigger a recreation of the pool, which is necessary to change container registry settings. In order to change the password for a container registry, either taint the Batch Pool and trigger an update, or manually destroy and re-create the pool.
```
If we don't read in the password each time updates to password should be reflected?

katbyte commented on azurerm/helpers/azure/batch_pool.go:
```diff
@@ -203,10 +204,44 @@ func FlattenBatchPoolContainerConfiguration(armContainerConfiguration *batch.Con
+	if userName := armContainerRegistry.UserName; userName != nil {
+		result["user_name"] = *userName
+	}
+	// We won't usually get the password back, but we include it here anyways
```
If this is usually not returned we should not be setting it. What you can do is re use the password from state that you could get via `d.Get()`

ennui93 commented:
Hi @katbyte, thanks again for your second review. I believe I have addressed all of your comments.

You are correct that I was unaware how to pass the password supplied by the user into the state. The solution looks much better now.

Please review at your convenience.
